sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala

```diff
@@ -48,3 +48,38 @@ case class Md5(child: Expression)
         s"(org.apache.commons.codec.digest.DigestUtils.md5Hex($c))")
   }
 }
+
+/**
+ * A function that calculates a sha1 hash value and returns it as a hex string
+ * For input of type [[BinaryType]] or [[StringType]]
+ */
+case class Sha1(child: Expression) extends UnaryExpression {
+
+  override def dataType: DataType = StringType
```

Comment:
Either `BinaryType` or `StringType`. 

`Sha1` could extend the class `ExpectsInputTypes` and override the method:
`override def expectedChildTypes: Seq[DataType] = Seq(BinaryType, StringType)`

Does this the expected type check? When I run execute the following test; it crashes because of `scala.MatchError` in Line [65](https://github.com/apache/spark/pull/6963/files#diff-25282ab1377a3d87999d1d0d7a8ec270R65).

All `BinaryMathexpression`s provide `Seq(DoubleType, DoubleType)` for `left` and `right`. What I am tring to say is that I don't know, how to express `BinaryType` OR `StringType`